project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/emf/EMFGModelState.java

```diff
@@ -0,0 +1,175 @@
+        if (!(state instanceof EMFGModelState)) {
+            throw new IllegalArgumentException("Argument must be a EMFGModelState");
+        }
+        return ((EMFGModelState) state);
+    }
+
+    public String getSourceURI() {
+        String sourceURI = ClientOptionsUtil.getSourceUri(getClientOptions())
+            .orElseThrow(() -> new GLSPServerException("No source URI given to load model!"));
+        return sourceURI;
+    }
+
+    @Override
+    public void sessionDisposed(ClientSession clientSession) {
+        closeResourceSet();
+    }
+
+    @SuppressWarnings("checkstyle:IllegalCatch")
+    protected void closeResourceSet() {
+        boolean result = false;
+        EList<Resource> resourceList = resourceSet.getResources();
+        for (Resource resource : resourceList) {
+            if (resource.getURI() != null) {
+                try {
+                    resource.unload();
+                    result = true;
+                } catch (RuntimeException e) {
+                    result = false;
+                    LOGGER.error("Could not unload resource: " + resource.getURI(), e);
+                    break;
+                }
+            }
+        }
+        if (result) {
+            commandStack.flush();
+            commandStack.saveIsDone();
+        }
+    }
+
+    protected <T extends EObject> T loadResource(final String sourceURI,
+        final Class<T> modelClass) throws IOException {
+
+        URI resourceURI = URI.createFileURI(sourceURI);
+        Optional<Resource> loadedResource = Optional.ofNullable(this.resourceSet.getResource(resourceURI, false))
+            .filter(Resource::isLoaded);
+
+        Resource resource = null;
+        if (loadedResource.isPresent()) {
+            resource = loadedResource.get();
+        } else {
+            resource = this.resourceSet.getResource(resourceURI, true);
+            resource.load(Collections.EMPTY_MAP);
+        }
+
+        EcoreUtil.resolveAll(resource);
+        EObject root = resource.getContents().get(0);
+        if (!modelClass.isInstance(root)) {
+            LOGGER.info("Resource appears to be empty: " + sourceURI);
+            return null;
+        }
+        return modelClass.cast(root);
+    }
+
+    public void saveResources() {
+        boolean result = false;
+        EList<Resource> resourceList = this.resourceSet.getResources();
+        for (Resource resource : resourceList) {
+            if (resource.getURI() != null) {
+                try {
+                    resource.save(Collections.EMPTY_MAP);
+                    result = true;
+                } catch (IOException e) {
+                    LOGGER.error("Could not save resource: " + resource.getURI(), e);
+                    throw new GLSPServerException("Could not save model to file: " + resource.getURI(), e);
+                }
+            }
+        }
+        if (result) {
+            this.commandStack.saveIsDone();
+        }
+    }
```

Comment:
I would put this code into an `EMFSourceModelStorage`.

Reply:
It doesn't show nicely anymore in Github. I meant here the saveResource and loadResource methods.
In general my goal would be that the model state is not concerned with loading or saving at all, but just holds the resource set (and closes it on dispose).

The source model loader should be responsible for loading and saving.
Also it is responsible to parse which URI to load from the arguments, etc.

Reply:
Ah thanks, I overlooked that in the first run I guess. With the current commit this should be resolved and the `SourceModelStorage` is responsible for the resource handling. The ResourceSet is stored in the ModelState.